### PR TITLE
Replace react chat engine to use native reactive agent

### PR DESCRIPTION
**Pull request by Disiok**
# Description

Replace react chat engine to use native reactive agent
Set default chat mode to best: use openai when possible, otherwise react

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

### CODE REVIEW

File: llama_index/indices/base.py [hunk 1]
```diff
@@ -340,30 +342,65 @@ def as_query_engine(self, **kwargs: Any) -> BaseQueryEngine:
         return RetrieverQueryEngine.from_args(**kwargs)
 
     def as_chat_engine(
-        self, chat_mode: ChatMode = ChatMode.CONDENSE_QUESTION, **kwargs: Any
+        self, chat_mode: ChatMode = ChatMode.BEST, **kwargs: Any
     ) -> BaseChatEngine:
+        query_engine = self.as_query_engine(**kwargs)
+        if "service_context" not in kwargs:
```
Comment (logan-markewich) [inline llama_index/indices/base.py:348]: this is a good fix, people often before would pass it in and get errors about it being included twice

File: llama_index/indices/base.py [hunk 2]
```diff
@@ -340,30 +342,65 @@ def as_query_engine(self, **kwargs: Any) -> BaseQueryEngine:
...
-                kwargs["service_context"] = self._service_context
-            return ReActChatEngine.from_query_engine(
-                query_engine=query_engine,
-                **kwargs,
+            from llama_index.agent import OpenAIAgent, ReActAgent
+            from llama_index.tools.query_engine import QueryEngineTool
+
+            # build query engine tool
+            name = kwargs.pop("name", None)
+            description = kwargs.pop("description", None)
```
Comment (logan-markewich) [inline llama_index/indices/base.py:377]: are these kwargs documented?

**Review by logan-markewich — APPROVED** [review]
LGTM, just a few very minor comments

File: llama_index/indices/base.py (hunk 2 above)
Comment (Disiok) [reply llama_index/indices/base.py:377]: they are not, in hind sight this is way too hidden anyways. will remove.

**Pull request merged**